File: オムロンカメラセンサ/Sound.vb
Imports System.Windows.Forms
Imports HVCW.CameraControl

''' <summary >
''' Camera registration (generate and play pairing sound)
''' </summary >
Partial Public Class Sound
    Inherits Form
    Private txtToken As String
    Private clsHvcw As ClassHvcw
    Private player As System.Media.SoundPlayer = Nothing

    Public Sub New()
        InitializeComponent()
    End Sub

    ' ***** Dialog display *****
    Public Sub SetDialog(owner As IWin32Window, refHvcw As ClassHvcw, ssid As String, token As String)
        clsHvcw = refHvcw
        textSSID.Text = ssid
        txtToken = token
        Me.ShowDialog(owner)
    End Sub

    ' ***** Stop sound playing *****
    Private Sub StopSound()
        If player IsNot Nothing Then
            player.[Stop]()
            player.Dispose()
            player = Nothing
        End If
    End Sub

    ' ***** Start pairing (play sound)*****
    Private Sub button1_Click(sender As Object, e As EventArgs)
        If clsHvcw.GenerateSound(textSSID.Text, textPassword.Text, txtToken) = True Then
            ' Read sound file
            Dim buf As Byte() = System.IO.File.ReadAllBytes(clsHvcw.SoundFile)

            ' Stop when sound playing
            If player IsNot Nothing Then
                StopSound()
            End If

            player = New System.Media.SoundPlayer()

            ' Wav header definition
            Dim wavHdr As New WAVHDR()

            Dim fs As UInteger = 8000

            wavHdr.formatid = &H1
            ' PCM uncompressed
            wavHdr.channel = 1
            ' ch=1 mono
            wavHdr.fs = fs
            ' Frequency
            wavHdr.bytespersec = fs * 2
            ' 16bit
            wavHdr.blocksize = 2
            ' 16bit mono so block size (byte/sample x # of channels) is 2
            wavHdr.bitspersample = 16
            ' bit/sample
            wavHdr.size = CUInt(buf.Length)
            ' Wave data byte number
            wavHdr.fileSize = wavHdr.size + CUInt(Marshal.SizeOf(wavHdr))
            ' Total byte number
            ' Play sound through memory stream
            Dim memoryStream As New System.IO.MemoryStream(CInt(wavHdr.fileSize))
            Dim bWriter As New System.IO.BinaryWriter(memoryStream)

            ' Write Wav header
            For Each b As Byte In wavHdr.getByteArray()
                bWriter.Write(b)
            Next
            ' Write PCM data
            For Each data As Byte In buf
                bWriter.Write(data)
            Next
            bWriter.Flush()

            memoryStream.Seek(0, System.IO.SeekOrigin.Begin)
            player.Stream = memoryStream

            ' Async play

            ' Wait until sound playing is over with following:
            ' player.PlaySync();
            player.Play()
        Else
            MessageBox.Show("Pairing sound creation failed", "ERROR", MessageBoxButtons.OK, MessageBoxIcon.[Error])
        End If
    End Sub

    ''' <summary >
    ''' Wav header definition
    ''' </summary >
    <StructLayout(LayoutKind.Sequential)>
    Public Class WAVHDR
        <MarshalAs(UnmanagedType.I4)>
        Public riff As UInt32 = &H46464952
        ' RIFF 
        <MarshalAs(UnmanagedType.I4)>
        Public fileSize As UInt32
        <MarshalAs(UnmanagedType.I4)>
        Public wave As UInt32 = &H45564157
        ' WAVE 
        <MarshalAs(UnmanagedType.I4)>
        Public fmt As UInt32 = &H20746D66
        ' fmt  
        <MarshalAs(UnmanagedType.I4)>
        Public fmtbytes As UInt32 = 16
        <MarshalAs(UnmanagedType.I2)>
        Public formatid As UInt16
        <MarshalAs(UnmanagedType.I2)>
        Public channel As UInt16
        <MarshalAs(UnmanagedType.I4)>
        Public fs As UInt32
        <MarshalAs(UnmanagedType.I4)>
        Public bytespersec As UInt32
        <MarshalAs(UnmanagedType.I2)>
        Public blocksize As UInt16
        <MarshalAs(UnmanagedType.I2)>
        Public bitspersample As UInt16
        <MarshalAs(UnmanagedType.I4)>
        Public data As UInt32 = &H61746164
        ' data 
        <MarshalAs(UnmanagedType.I4)>
        Public size As UInt32

        ' convert the struct to byte array
        Public Function getByteArray() As Byte()
            Dim len As Integer = Marshal.SizeOf(Me)
            Dim arr As Byte() = New Byte(len - 1) {}
            Dim ptr As IntPtr = Marshal.AllocHGlobal(len)
            'or false
            Marshal.StructureToPtr(Me, ptr, True)
            'or arr.Length
            Marshal.Copy(ptr, arr, 0, len)
            Marshal.FreeHGlobal(ptr)
            Return arr
        End Function
    End Class
End Class

File: OmronCameraDemo/Sound.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using System.Runtime.InteropServices;

namespace OmronCameraDemo
{
    /// <summary >
    /// Camera registration (generate and play pairing sound)
    /// </summary >
    public partial class Sound : Form
    {
        string txtToken;
        ClassHvcw clsHvcw;
        System.Media.SoundPlayer player = null;

        public Sound()
        {
            InitializeComponent();
        }

        // ***** Dialog display *****
        public void SetDialog(IWin32Window owner, ClassHvcw refHvcw, string ssid, string token)
        {
            clsHvcw = refHvcw;
            textSSID.Text = ssid;
            txtToken = token;
            this.ShowDialog(owner);
        }

        // ***** Stop sound playing *****
        private void StopSound()
        {
            if (player != null)
            {
                player.Stop();
                player.Dispose();
                player = null;
            }
        }

        // ***** Start pairing (play sound)*****
        private void button1_Click(object sender, EventArgs e)
        {
            if (clsHvcw.GenerateSound(textSSID.Text, textPassword.Text, txtToken) == true)
            {
                // Read sound file
                byte[] buf = System.IO.File.ReadAllBytes(clsHvcw.SoundFile);

                // Stop when sound playing
                if (player != null)
                    StopSound();

                player = new System.Media.SoundPlayer();

                // Wav header definition
                WAVHDR wavHdr = new WAVHDR();

                uint fs = 8000;

                wavHdr.formatid = 0x0001;                                       // PCM uncompressed
                wavHdr.channel = 1;                                             // ch=1 mono
                wavHdr.fs = fs;                                                 // Frequency
                wavHdr.bytespersec = fs * 2;                                    // 16bit
                wavHdr.blocksize = 2;                                           // 16bit mono so block size (byte/sample x # of channels) is 2
                wavHdr.bitspersample = 16;                                      // bit/sample
                wavHdr.size = (uint)buf.Length;                                 // Wave data byte number
                wavHdr.fileSize = wavHdr.size + (uint)Marshal.SizeOf(wavHdr);   // Total byte number

                // Play sound through memory stream
                System.IO.MemoryStream memoryStream = new System.IO.MemoryStream((int)wavHdr.fileSize);
                System.IO.BinaryWriter bWriter = new System.IO.BinaryWriter(memoryStream);

                // Write Wav header
                foreach (byte b in wavHdr.getByteArray())
                {
                    bWriter.Write(b);
                }
                // Write PCM data
                foreach (byte data in buf)
                {
                    bWriter.Write(data);
                }
                bWriter.Flush();

                memoryStream.Seek(0, System.IO.SeekOrigin.Begin);
                player.Stream = memoryStream;

                // Async play
                player.Play();

                // Wait until sound playing is over with following:
                // player.PlaySync();
            }
            else
            {
                MessageBox.Show("Pairing sound creation failed", "ERROR", MessageBoxButtons.OK, MessageBoxIcon.Error);
            }
        }

        /// <summary >
        /// Wav header definition
        /// </summary >
        [StructLayout(LayoutKind.Sequential)]
        public class WAVHDR
        {
            [MarshalAs(UnmanagedType.I4)]
            public UInt32 riff = 0x46464952; /* RIFF */
            [MarshalAs(UnmanagedType.I4)]
            public UInt32 fileSize;
            [MarshalAs(UnmanagedType.I4)]
            public UInt32 wave = 0x45564157; /* WAVE */
            [MarshalAs(UnmanagedType.I4)]
            public UInt32 fmt = 0x20746D66; /* fmt  */
            [MarshalAs(UnmanagedType.I4)]
            public UInt32 fmtbytes = 16;
            [MarshalAs(UnmanagedType.I2)]
            public UInt16 formatid;
            [MarshalAs(UnmanagedType.I2)]
            public UInt16 channel;
            [MarshalAs(UnmanagedType.I4)]
            public UInt32 fs;
            [MarshalAs(UnmanagedType.I4)]
            public UInt32 bytespersec;
            [MarshalAs(UnmanagedType.I2)]
            public UInt16 blocksize;
            [MarshalAs(UnmanagedType.I2)]
            public UInt16 bitspersample;
            [MarshalAs(UnmanagedType.I4)]
            public UInt32 data = 0x61746164; /* data */
            [MarshalAs(UnmanagedType.I4)]
            public UInt32 size;

            // convert the struct to byte array
            public byte[] getByteArray()
            {
                int len = Marshal.SizeOf(this);
                byte[] arr = new byte[len];
                IntPtr ptr = Marshal.AllocHGlobal(len);
                Marshal.StructureToPtr(this, ptr, true/*or false*/);
                Marshal.Copy(ptr, arr, 0, len/*or arr.Length*/);
                Marshal.FreeHGlobal(ptr);
                return arr;
            }
        }
    }
}
